util/automation.py

```diff
@@ -24,6 +39,139 @@
+                    assert len(d) > 1
+                    writer.writerow(self.merge_line(d))
+                    f.flush()
+
+    def ingest(self, line):
+        return map(lambda x: x.replace('-', '0'), filter(lambda x: len(x) > 0, map(lambda x: x.strip(), line.split(' ')[1:])))
+
+    def merge_line(self, line):
+        return dict(zip(self._names, line))
+
+class GPUUsageChart():
+    def __init__(self, source, basename=None):
+        self._rows    = [ 'pwr', 'temp', 'sm', 'mem']
+        self._titles  = {
+            'pwr':  "Power (W)",
+            'temp': "Temperature (°C)",
+            'sm':   "Streaming Multiprocessors (%)",
+            'mem':  "Memory (%)"
+        }
+        self._data     = { }.fromkeys(self._rows)
+        self._csv      = source
+        self._basename = basename or os.environ.get('ds_gpu_usage_charts', 'gpu_usage_%%s_%d.png' % int(time.time()))
+
+        # This should make sure we start from anything clean.
+        plt.close("all")
+
+        try:
+            self.read()
+            for plot in self._rows:
+                self.produce_plot(plot)
+        except IOError as ex:
+            print "Unable to read", ex
+
+    def append_data(self, row):
+        for bucket, value in row.iteritems():
+            if not bucket in self._rows:
+                continue
+
+            if not self._data[bucket]:
+                self._data[bucket] = {}
+
+            gpu = int(row['gpu'])
+            if not self._data[bucket].has_key(gpu):
+                self._data[bucket][gpu]  = [ value ]
+            else:
+                self._data[bucket][gpu] += [ value ]
+
+    def read(self):
+        print "Reading data from", self._csv
+        with open(self._csv, 'r') as f:
+            for r in csv.DictReader(f):
+                self.append_data(r)
+
+    def produce_plot(self, key, with_spline=True):
+        png = self._basename % (key, )
+        print "Producing plot for", key, "as", png
+        fig, axis = plt.subplots()
+        data = self._data[key]
+        if data is None:
+            print "Data was empty, aborting"
+            return
+
+        x = range(len(data[0]))
+        if with_spline:
+            x = map(lambda x: float(x), x)
+            x_sm = np.array(x)
+            x_smooth = np.linspace(x_sm.min(), x_sm.max(), 300)
+
+        for gpu, y in data.iteritems():
+            if with_spline:
+                y = map(lambda x: float(x), y)
+                y_sm = np.array(y)
+                y_smooth = spline(x, y, x_smooth, order=1)
+                axis.plot(x_smooth, y_smooth, label='GPU %d' % (gpu))
+            else:
+                axis.plot(x, y, label='GPU %d' % (gpu))
+
+        axis.legend(loc="upper right", frameon=False)
+        axis.set_xlabel("Time (s)")
+        axis.set_ylabel("%s" % self._titles[key])
+        fig.set_size_inches(24, 18)
+        plt.title("GPU Usage: %s" % self._titles[key])
+        plt.savefig(png, dpi=100)
+        plt.close(fig)
```

Comment:
Required otherwise we leak figures and matplotlib complains (and system memory too)